src/query/service/src/interpreters/interpreter_table_analyze.rs

```diff
@@ -144,46 +181,163 @@ impl Interpreter for AnalyzeTableInterpreter {
-            let (plan, _) = planner.plan_sql(&sql).await?;
-            let (select_plan, schema) = match &plan {
-                Plan::Query {
-                    s_expr,
-                    metadata,
-                    bind_context,
-                    ..
-                } => {
-                    let mut builder1 =
-                        PhysicalPlanBuilder::new(metadata.clone(), self.ctx.clone(), false);
-                    (
-                        builder1.build(s_expr, bind_context.column_set()).await?,
-                        bind_context.output_schema(),
-                    )
-                }
-                _ => unreachable!(),
-            };
+            info!("Analyze via sql {:?}", sql);
 
+            let (physical_plan, bind_context) = self.plan_sql(sql).await?;
             let mut build_res =
-                build_query_pipeline_without_render_result_set(&self.ctx, &select_plan).await?;
+                build_query_pipeline_without_render_result_set(&self.ctx, &physical_plan).await?;
+            // After profiling, computing histogram is heavy and the bottleneck is window function(90%).
+            // It's possible to OOM if the table is too large and spilling isn't enabled.
+            // We add a setting `enable_analyze_histogram` to control whether to compute histogram(default is closed).
+            let mut histogram_info_receivers = HashMap::new();
+            if self.ctx.get_settings().get_enable_analyze_histogram()? {
+                let histogram_sqls = index_cols
+                    .iter()
+                    .map(|c| {
+                        format!(
+                            "SELECT quantile,
+                            COUNT(DISTINCT {}) AS ndv,
+                            MAX({}) AS max_value,
+                            MIN({}) AS min_value,
+                            COUNT() as count
+                        FROM  (
+                            SELECT {}, NTILE({}) OVER (ORDER BY {}) AS quantile
+                            FROM {}.{} WHERE {} IS DISTINCT FROM NULL
+                        )
+                        GROUP BY quantile ORDER BY quantile \n",
+                            c.1,
+                            c.1,
+                            c.1,
+                            c.1,
+                            DEFAULT_HISTOGRAM_BUCKETS,
+                            c.1,
+                            plan.database,
+                            plan.table,
+                            c.1,
+                        )
+                    })
+                    .collect::<Vec<_>>();
+                for (sql, (col_id, _)) in histogram_sqls.into_iter().zip(index_cols.iter()) {
+                    info!("Analyze histogram via sql {:?}", sql);
+                    let (mut histogram_plan, bind_context) = self.plan_sql(sql).await?;
+                    if !self.ctx.get_cluster().is_empty() {
+                        histogram_plan = remove_exchange(histogram_plan);
```

Comment:
why do we need remove_exchange recursively?

Reply:
The exchange manager does not yet support the ability to build multiple pipelines for the same query ctx.